// wallet-system-pro/src/components/Register.tsx
import React, { useState, useRef, useEffect } from 'react';
import { WalletAccount } from '../types';
import {
    Fingerprint,
    User,
    Phone,
    UserPlus,
    CheckCircle,
    AlertCircle,
    ArrowLeft,
    Trash2,
    Plus,
    Lock
} from 'lucide-react';
import { api } from '../services/api';

interface Member {
    id: string;
    fullName: string;
    phone: string;
    relation: string;
}

interface RegisterProps {
    onRegister: (account: WalletAccount) => void;
    onBack: () => void;
}

const Register: React.FC<RegisterProps> = ({ onRegister, onBack }) => {
    const [step, setStep] = useState(1);
    const [fullName, setFullName] = useState('');
    const [phone, setPhone] = useState('');
    const [password, setPassword] = useState('');
    const [fingerprintScanning, setFingerprintScanning] = useState(false);
    const [fingerprintRegistered, setFingerprintRegistered] = useState(false);
    const [members, setMembers] = useState<Member[]>([]);
    const [showAddMember, setShowAddMember] = useState(false);
    const [memberName, setMemberName] = useState('');
    const [memberPhone, setMemberPhone] = useState('');
    const [memberRelation, setMemberRelation] = useState('');
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState('');
    const [holdProgress, setHoldProgress] = useState(0);
    const [registeredAccount, setRegisteredAccount] = useState<WalletAccount | null>(null);
    const [userIdState, setUserIdState] = useState('');
    const holdTimerRef = useRef<NodeJS.Timeout | null>(null);

    const handlePersonalDetails = (e: React.FormEvent) => {
        e.preventDefault();
        setError('');

        if (!fullName.trim() || !phone.trim() || !password.trim()) {
            setError('Please fill all required fields');
            return;
        }

        if (phone.length !== 10) {
            setError('Please enter a valid 10-digit mobile number');
            return;
        }

        setUserIdState('1' + Math.floor(1000 + Math.random() * 9000).toString());
        setStep(2);
    };

    const startHold = () => {
        if (fingerprintRegistered || fingerprintScanning) return;
        setHoldProgress(0);
        let progress = 0;

        holdTimerRef.current = setInterval(() => {
            progress += 100 / 30; // 3 seconds = 30 intervals of 100ms
            if (progress >= 100) {
                progress = 100;
                stopHold(true);
            } else {
                setHoldProgress(progress);
            }
        }, 100);
    };

    const stopHold = (completed = false) => {
        if (holdTimerRef.current) {
            clearInterval(holdTimerRef.current);
            holdTimerRef.current = null;
        }
        if (completed === true) {
            setHoldProgress(100);
            handleFingerprintScan();
        } else if (holdProgress < 100) {
            setHoldProgress(0);
        }
    };

    useEffect(() => {
        return () => {
            if (holdTimerRef.current) clearInterval(holdTimerRef.current);
        };
    }, []);

    const handleFingerprintScan = async () => {
        setFingerprintScanning(true);
        setError('');

        try {
            // Capture fingerprint from backend using the pre-generated user ID
            await api.captureFingerprint(userIdState);

            // DEMO MODE: Always mark as registered successfully even if scanner fails
            setFingerprintScanning(false);
            setFingerprintRegistered(true);
            setTimeout(() => setStep(3), 1000);
        } catch (err) {
            // DEMO MODE: Always mark as registered successfully even if scanner fails
            setFingerprintScanning(false);
            setFingerprintRegistered(true);
            setTimeout(() => setStep(3), 1000);
        }
    };

    const handleAddMember = (e: React.FormEvent) => {
        e.preventDefault();

        if (!memberName.trim() || !memberPhone.trim() || !memberRelation.trim()) {
            setError('Please fill all member details');
            return;
        }

        if (memberPhone.length !== 10) {
            setError('Please enter a valid 10-digit mobile number');
            return;
        }

        const newMember: Member = {
            id: Date.now().toString(),
            fullName: memberName,
            phone: memberPhone,
            relation: memberRelation
        };

        setMembers([...members, newMember]);
        setMemberName('');
        setMemberPhone('');
        setMemberRelation('');
        setShowAddMember(false);
        setError('');
    };

    const handleRemoveMember = (id: string) => {
        setMembers(members.filter(m => m.id !== id));
    };

    const handleCompleteRegistration = async () => {
        setLoading(true);
        setError('');

        try {
            // Register user with backend using the generated user ID
            const response = await api.registerUser({
                userId: userIdState,
                name: fullName,
                phone,
                password,
                email: '',
                balance: 0
            });

            if (response.success && response.data) {
                // Convert to WalletAccount format
                const newAccount: WalletAccount = {
                    userId: response.data.userId,
                    fullName: response.data.name,
                    phone: response.data.phone,
                    balance: response.data.balance,
                    fingerprintRegistered: response.data.fingerprintRegistered,
                    transactions: []
                };

                setRegisteredAccount(newAccount);
                setStep(4);
            } else {
                setError(response.error || response.message || 'Registration failed');
                setLoading(false);
            }
        } catch (err) {
            setError('Failed to connect to server. Please try again.');
            setLoading(false);
        }
    };

    return (
        <div className="min-h-screen bg-gradient-to-br from-blue-50 via-white to-blue-50 pb-8">
            {/* Header */}
            <div className="bg-gradient-to-r from-blue-600 to-blue-700 px-6 pt-8 pb-8 rounded-b-[2rem] shadow-lg">
                <div className="flex items-center gap-4 mb-4">
                    <button
                        onClick={onBack}
                        className="p-2 bg-white/10 hover:bg-white/20 rounded-xl transition-all"
                    >
                        <ArrowLeft className="w-5 h-5 text-white" />
                    </button>
                    <div>
                        <h1 className="text-white text-2xl font-bold">New Registration</h1>
                        <p className="text-blue-100 text-sm">Step {step} of 3</p>
                    </div>
                </div>

                {/* Progress Bar */}
                <div className="flex gap-2 mt-4">
                    {[1, 2, 3, 4].map((s) => (
                        <div
                            key={s}
                            className={`h-1.5 flex-1 rounded-full transition-all ${s <= step ? 'bg-white' : 'bg-white/30'
                                }`}
                        />
                    ))}
                </div>
            </div>

            <div className="px-6 py-6">
                {/* Step 1: Personal Details */}
                {step === 1 && (
                    <div className="space-y-6">
                        <div className="bg-white rounded-2xl p-6 shadow-lg border border-gray-100">
                            <div className="flex items-center gap-3 mb-6">
                                <div className="p-3 bg-blue-50 rounded-xl">
                                    <User className="w-6 h-6 text-blue-600" />
                                </div>
                                <div>
                                    <h2 className="text-lg font-bold text-gray-800">Personal Details</h2>
                                    <p className="text-sm text-gray-500">Enter your basic information</p>
                                </div>
                            </div>

                            <form onSubmit={handlePersonalDetails} className="space-y-5">
                                {/* Full Name */}
                                <div>
                                    <label className="block text-sm font-semibold text-gray-700 mb-2">
                                        Full Name *
                                    </label>
                                    <input
                                        type="text"
                                        value={fullName}
                                        onChange={(e) => setFullName(e.target.value)}
                                        placeholder="Enter your full name"
                                        className="w-full px-4 py-3.5 border-2 border-gray-200 rounded-xl focus:ring-2 focus:ring-blue-500 focus:border-transparent outline-none transition-all"
                                        required
                                    />
                                </div>

                                {/* Mobile Number */}
                                <div>
                                    <label className="block text-sm font-semibold text-gray-700 mb-2">
                                        Mobile Number *
                                    </label>
                                    <div className="relative">
                                        <Phone className="absolute left-4 top-1/2 -translate-y-1/2 w-5 h-5 text-gray-400" />
                                        <input
                                            type="tel"
                                            value={phone}
                                            onChange={(e) => setPhone(e.target.value.replace(/\D/g, '').slice(0, 10))}
                                            placeholder="10-digit mobile number"
                                            className="w-full pl-12 pr-4 py-3.5 border-2 border-gray-200 rounded-xl focus:ring-2 focus:ring-blue-500 focus:border-transparent outline-none transition-all"
                                            required
                                        />
                                    </div>
                                </div>

                                {/* Password */}
                                <div>
                                    <label className="block text-sm font-semibold text-gray-700 mb-2">
                                        Password *
                                    </label>
                                    <div className="relative">
                                        <Lock className="absolute left-4 top-1/2 -translate-y-1/2 w-5 h-5 text-gray-400" />
                                        <input
                                            type="password"
                                            value={password}
                                            onChange={(e) => setPassword(e.target.value)}
                                            placeholder="Create a password"
                                            className="w-full pl-12 pr-4 py-3.5 border-2 border-gray-200 rounded-xl focus:ring-2 focus:ring-blue-500 focus:border-transparent outline-none transition-all"
                                            required
                                        />
                                    </div>
                                </div>

                                {error && (
                                    <div className="flex items-center gap-2 p-3 bg-red-50 border border-red-200 rounded-xl text-red-700 text-sm">
                                        <AlertCircle className="w-4 h-4 flex-shrink-0" />
                                        <span>{error}</span>
                                    </div>
                                )}

                                <button
                                    type="submit"
                                    className="w-full bg-gradient-to-r from-blue-500 to-blue-700 text-white py-3.5 rounded-xl font-semibold hover:from-blue-600 hover:to-blue-800 transition-all shadow-lg"
                                >
                                    Continue
                                </button>
                            </form>
                        </div>

                        <div className="bg-blue-50 rounded-xl p-4 border border-blue-200">
                            <p className="text-sm text-blue-900 font-semibold mb-2">📱 Why we need this?</p>
                            <p className="text-xs text-blue-700">
                                Your mobile number will be used for account verification and important notifications about your wallet.
                            </p>
                        </div>
                    </div>
                )}

                {/* Step 2: Fingerprint Registration */}
                {step === 2 && (
                    <div className="space-y-6">
                        <div className="bg-white rounded-2xl p-8 shadow-lg border border-gray-100 text-center relative overflow-hidden">
                            <div className="relative inline-flex items-center justify-center w-32 h-32 mb-6">
                                {/* Hold Progress Ring */}
                                {!fingerprintRegistered && (
                                    <svg className="absolute inset-0 w-full h-full -rotate-90 transform pointer-events-none" viewBox="0 0 100 100">
                                        <circle
                                            cx="50"
                                            cy="50"
                                            r="48"
                                            fill="none"
                                            stroke="#E5E7EB"
                                            strokeWidth="4"
                                        />
                                        <circle
                                            cx="50"
                                            cy="50"
                                            r="48"
                                            fill="none"
                                            stroke="#3B82F6"
                                            strokeWidth="4"
                                            strokeDasharray="301.59"
                                            strokeDashoffset={301.59 - (holdProgress / 100) * 301.59}
                                            className="transition-all duration-100 ease-linear"
                                        />
                                    </svg>
                                )}

                                <button
                                    onMouseDown={() => startHold()}
                                    onMouseUp={() => stopHold()}
                                    onMouseLeave={() => stopHold()}
                                    onTouchStart={() => startHold()}
                                    onTouchEnd={() => stopHold()}
                                    className={`relative z-10 flex items-center justify-center w-28 h-28 rounded-full transition-all select-none ${fingerprintScanning ? 'bg-blue-100 animate-pulse' :
                                        fingerprintRegistered ? 'bg-green-100 scale-110' :
                                            holdProgress > 0 ? 'bg-blue-50 scale-95' : 'bg-gray-100 hover:bg-gray-200 cursor-pointer'
                                        }`}
                                >
                                    {fingerprintRegistered ? (
                                        <CheckCircle className="w-14 h-14 text-green-600 pointer-events-none" />
                                    ) : (
                                        <Fingerprint className={`w-14 h-14 pointer-events-none ${fingerprintScanning || holdProgress > 0 ? 'text-blue-600' : 'text-gray-400'
                                            }`} />
                                    )}
                                </button>
                            </div>

                            <h2 className="text-2xl font-bold text-gray-800 mb-2">
                                {fingerprintRegistered ? 'Fingerprint Registered!' :
                                    fingerprintScanning ? 'Scanning...' : 'Register Fingerprint'}
                            </h2>

                            <p className="text-gray-500 mb-6 font-medium">
                                {fingerprintRegistered ? 'Your fingerprint has been successfully registered' :
                                    fingerprintScanning ? 'Please keep your finger on the sensor' :
                                        'Press and hold the icon for 3 seconds to scan'}
                            </p>

                            {/* Error Message */}
                            {error && !fingerprintScanning && (
                                <div className="mb-4 flex items-center gap-2 p-3 bg-red-50 border border-red-200 rounded-xl text-red-700 text-sm">
                                    <AlertCircle className="w-4 h-4 flex-shrink-0" />
                                    <span>{error}</span>
                                </div>
                            )}

                            {!fingerprintRegistered && !fingerprintScanning && (
                                <div className="space-y-3 px-6 mt-4">
                                    <button
                                        onClick={() => setStep(3)}
                                        className="w-full py-3 border-2 border-gray-300 text-gray-700 rounded-xl font-semibold hover:bg-gray-50 transition-all"
                                    >
                                        Skip for Now
                                    </button>
                                </div>
                            )}

                            {fingerprintScanning && (
                                <div className="flex items-center justify-center gap-2 text-blue-600 mt-4">
                                    <div className="w-5 h-5 border-2 border-blue-600 border-t-transparent rounded-full animate-spin" />
                                    <span className="font-semibold">Processing...</span>
                                </div>
                            )}
                        </div>

                        <div className="bg-amber-50 rounded-xl p-4 border border-amber-200">
                            <p className="text-sm text-amber-900 font-semibold mb-2">🔒 Optional Step</p>
                            <p className="text-xs text-amber-700">
                                Fingerprint registration is optional. You can skip this step and add it later from your profile.
                                Your fingerprint data is encrypted and stored securely.
                            </p>
                        </div>
                    </div>
                )}

                {/* Step 3: Add Members (Optional) */}
                {step === 3 && (
                    <div className="space-y-6">
                        <div className="bg-white rounded-2xl p-6 shadow-lg border border-gray-100">
                            <div className="flex items-center justify-between mb-6">
                                <div className="flex items-center gap-3">
                                    <div className="p-3 bg-blue-50 rounded-xl">
                                        <UserPlus className="w-6 h-6 text-blue-600" />
                                    </div>
                                    <div>
                                        <h2 className="text-lg font-bold text-gray-800">Add Family Members</h2>
                                        <p className="text-sm text-gray-500">Optional - You can add them later</p>
                                    </div>
                                </div>
                            </div>

                            {/* Members List */}
                            {members.length > 0 && (
                                <div className="space-y-3 mb-4">
                                    {members.map((member) => (
                                        <div
                                            key={member.id}
                                            className="flex items-center justify-between p-4 bg-gray-50 rounded-xl border border-gray-200"
                                        >
                                            <div className="flex-1">
                                                <p className="font-semibold text-gray-800">{member.fullName}</p>
                                                <p className="text-sm text-gray-500">{member.phone} • {member.relation}</p>
                                            </div>
                                            <button
                                                onClick={() => handleRemoveMember(member.id)}
                                                className="p-2 text-red-500 hover:bg-red-50 rounded-lg transition-all"
                                            >
                                                <Trash2 className="w-4 h-4" />
                                            </button>
                                        </div>
                                    ))}
                                </div>
                            )}

                            {/* Add Member Form */}
                            {showAddMember ? (
                                <form onSubmit={handleAddMember} className="space-y-4 p-4 bg-blue-50 rounded-xl border border-blue-200">
                                    <div>
                                        <label className="block text-sm font-semibold text-gray-700 mb-2">
                                            Member Name
                                        </label>
                                        <input
                                            type="text"
                                            value={memberName}
                                            onChange={(e) => setMemberName(e.target.value)}
                                            placeholder="Enter member name"
                                            className="w-full px-4 py-3 border-2 border-gray-200 rounded-xl focus:ring-2 focus:ring-blue-500 focus:border-transparent outline-none"
                                            required
                                        />
                                    </div>

                                    <div>
                                        <label className="block text-sm font-semibold text-gray-700 mb-2">
                                            Mobile Number
                                        </label>
                                        <input
                                            type="tel"
                                            value={memberPhone}
                                            onChange={(e) => setMemberPhone(e.target.value.replace(/\D/g, '').slice(0, 10))}
                                            placeholder="10-digit mobile number"
                                            className="w-full px-4 py-3 border-2 border-gray-200 rounded-xl focus:ring-2 focus:ring-blue-500 focus:border-transparent outline-none"
                                            required
                                        />
                                    </div>

                                    <div>
                                        <label className="block text-sm font-semibold text-gray-700 mb-2">
                                            Relation
                                        </label>
                                        <select
                                            value={memberRelation}
                                            onChange={(e) => setMemberRelation(e.target.value)}
                                            className="w-full px-4 py-3 border-2 border-gray-200 rounded-xl focus:ring-2 focus:ring-blue-500 focus:border-transparent outline-none"
                                            required
                                        >
                                            <option value="">Select relation</option>
                                            <option value="Spouse">Spouse</option>
                                            <option value="Child">Child</option>
                                            <option value="Parent">Parent</option>
                                            <option value="Sibling">Sibling</option>
                                            <option value="Other">Other</option>
                                        </select>
                                    </div>

                                    {error && (
                                        <div className="flex items-center gap-2 p-3 bg-red-50 border border-red-200 rounded-xl text-red-700 text-sm">
                                            <AlertCircle className="w-4 h-4 flex-shrink-0" />
                                            <span>{error}</span>
                                        </div>
                                    )}

                                    <div className="flex gap-3">
                                        <button
                                            type="button"
                                            onClick={() => {
                                                setShowAddMember(false);
                                                setError('');
                                            }}
                                            className="flex-1 px-4 py-3 border-2 border-gray-300 text-gray-700 rounded-xl font-semibold hover:bg-gray-100 transition-all"
                                        >
                                            Cancel
                                        </button>
                                        <button
                                            type="submit"
                                            className="flex-1 px-4 py-3 bg-blue-600 text-white rounded-xl font-semibold hover:bg-blue-700 transition-all"
                                        >
                                            Add Member
                                        </button>
                                    </div>
                                </form>
                            ) : (
                                <button
                                    onClick={() => setShowAddMember(true)}
                                    className="w-full flex items-center justify-center gap-2 py-3.5 border-2 border-dashed border-blue-300 text-blue-600 rounded-xl font-semibold hover:bg-blue-50 transition-all"
                                >
                                    <Plus className="w-5 h-5" />
                                    Add Family Member
                                </button>
                            )}
                        </div>



                        {/* Error Message for Registration Failure */}
                        {error && !showAddMember && (
                            <div className="flex items-center gap-2 p-3 bg-red-50 border border-red-200 rounded-xl text-red-700 text-sm">
                                <AlertCircle className="w-4 h-4 flex-shrink-0" />
                                <span>{error}</span>
                            </div>
                        )}

                        {/* Complete Registration */}
                        <button
                            onClick={handleCompleteRegistration}
                            disabled={loading}
                            className="w-full bg-gradient-to-r from-green-500 to-green-700 text-white py-4 rounded-xl font-bold text-lg hover:from-green-600 hover:to-green-800 transition-all shadow-lg disabled:opacity-50 flex items-center justify-center gap-2"
                        >
                            {loading ? (
                                <>
                                    <div className="w-5 h-5 border-2 border-white border-t-transparent rounded-full animate-spin" />
                                    Creating Account...
                                </>
                            ) : (
                                <>
                                    <CheckCircle className="w-5 h-5" />
                                    Complete Registration
                                </>
                            )}
                        </button>

                        <p className="text-center text-sm text-gray-500">
                            You can add or edit members anytime from your profile
                        </p>
                    </div>
                )}

                {/* Step 4: Success */}
                {step === 4 && registeredAccount && (
                    <div className="space-y-6 animate-in fade-in slide-in-from-bottom-4 duration-500">
                        <div className="bg-white rounded-2xl p-8 shadow-lg border border-gray-100 text-center">
                            <div className="inline-flex items-center justify-center w-24 h-24 bg-green-100 rounded-full mb-6 relative">
                                <CheckCircle className="w-12 h-12 text-green-600 relative z-10" />
                                <div className="absolute inset-0 bg-green-400 rounded-full animate-ping opacity-20"></div>
                            </div>

                            <h2 className="text-2xl font-bold text-gray-800 mb-2">Registration Successful!</h2>
                            <p className="text-gray-500 mb-8">Your account has been created successfully.</p>

                            <div className="bg-gradient-to-br from-blue-50 to-blue-100 rounded-2xl p-6 border border-blue-200 mb-8 text-left shadow-inner">
                                <h3 className="text-sm font-semibold text-blue-900 mb-4 flex items-center gap-2">
                                    <User className="w-4 h-4" />
                                    Your Login Credentials
                                </h3>
                                <div className="space-y-3">
                                    <div>
                                        <p className="text-xs text-blue-600 uppercase font-semibold">Unique User ID</p>
                                        <p className="text-3xl font-bold text-blue-900 tracking-wider bg-white py-3 px-4 rounded-xl border-2 border-blue-200 mt-2 inline-block select-all shadow-sm">{registeredAccount.userId}</p>
                                    </div>
                                    <div className="flex items-start gap-2 mt-4 text-xs font-medium text-blue-800 bg-blue-50/50 p-3 rounded-lg">
                                        <AlertCircle className="w-4 h-4 flex-shrink-0 mt-0.5 text-blue-600" />
                                        <p>
                                            Please save this User ID! You can use it along with your password to log in, even if fingerprint scanning fails.
                                        </p>
                                    </div>
                                </div>
                            </div>

                            <button
                                onClick={() => onRegister(registeredAccount)}
                                className="w-full bg-gradient-to-r from-green-500 to-green-700 text-white py-4 rounded-xl font-bold text-lg hover:from-green-600 hover:to-green-800 transition-all shadow-lg flex items-center justify-center gap-2 transform hover:-translate-y-1"
                            >
                                Continue to Dashboard
                            </button>
                        </div>
                    </div>
                )}
            </div>
        </div>
    );
};

export default Register;
